### DISCUSSION TITLE
PHP 8.1

### DISCUSSION THREAD

**Discussion by soulm69** [PHPOffice/PhpSpreadsheet#3242] [Q&A] [answered]
Hi PhpSpreadsheet Team,

May I know when will PHP 8.1 be supported?

Thank you.

**Comment by oleibman** [comment]
Php8.1 has been supported for some time now, and Php8.2 is now supported as well. Are you having a particular problem?

**Reply by soulm69** [reply]
Hi oleibman,

I have a web project that creates XLSX files.

On PHP 8.0 (64bit), it is working perfectly. But on PHP 8.1 (64bit), I can't open the resulting XLSX file, says the file is corrupted or something.

I have PhpSpreadsheet 1.25.2, installed it using Composer.

I use $xlsWriter = IOFactory::createWriter($xls, 'Xlsx') and $xlsWriter->save($fileName).

**Comment by MarkBaker** [comment]
We'd need a bit more information about the actual content of that file: we run an extensive set of tests against all supported versions of PHP, including 8.1, 8.2 and the nightly build for every commit or PR to the library; these include tests for the `save()` method. It's possible that there is some feature that isn't fully covered by those tests, but we can't guess what that might be.

So what is the content and structure of your file; are you using features like headers/footers, conditional formatting, data validations, tables, etc?

Does MS Excel try to fix the file when it is opened? (that's normal behaviour if there is a problem with Excel file content) If so, what messages does it display?

Is anything being written to your logs files?

**Comment by soulm69** [comment, marked as answer]
Sorry, it's all good now.

I had a different error not related to PhpSpreadsheet.

Thank you!

**Comment by ZakariaBghr** [comment]
hey people, can anybody give me a site or sth where I can see some php code examples ?

**Reply by ZakariaBghr** [reply]
for beginners

**Reply by MarkBaker** [reply]
> hey people, can anybody give me a site or sth where I can see some php code examples ?

Have you looked at the [examples](https://github.com/PHPOffice/PhpSpreadsheet/tree/master/samples) and the [documentation](https://phpspreadsheet.readthedocs.io/en/latest/)

What are you finding difficult?

**Reply by MarkBaker** [reply]
> for beginners

And if you need to ask something, it's always a good idea to raise it as a new question rather than to piggy-back your question into something completely unrelated

**Reply by ZakariaBghr** [reply]
no I haven't

**Reply by ZakariaBghr** [reply]
![Screenshot_20230214_072606](https://user-images.githubusercontent.com/122107884/218825278-abf23845-346e-4694-b36f-68e2f7a05d45.png)

**Reply by ZakariaBghr** [reply]
I don't why it doesn't remove the elements from the array

**Reply by MarkBaker** [reply]
> I don't why it doesn't remove the elements from the array

That seems to be a very general PHP question, nothing whatsoever to do with PhpSpreadsheet, so I don't really know why you've chosen to ask it here.

But in answer, it is removing elements from the array... the problem is that you're removing it from the array that's **_inside_** the `card_deal()` function; that won't change anything **_outside_** that function.
If you want the `$tab` array to be changed in your calling function, then you have to declare it call by reference in your function definition.

```php
function card_deal(array &$tab) {
    $ran = rand(0, count($tab) - 1);
    $card = $tab[$ran];
    array_splice($tab, $ran, 1);
    return $card;
}
```
Note the `&` in the function definition.

P.S. Posting images of code is extremely unhelpful: I can't test/run an image of code, I have to write it out by hand, which is creating work and using more of my time to answer your question. Couldn't you have used copy/paste from your IDE? You can wrap code examples in triple backticks here, because we use markdown.

**Reply by ZakariaBghr** [reply]
okay thank you sir, I'm sorry as you might have noticed I'm new to github and I haven't even noticed the topic.
Sorry for the inconvenience

**Reply by MarkBaker** [reply]
> okay thank you sir, I'm sorry as you might have noticed I'm new to github and I haven't even noticed the topic.
> Sorry for the inconvenience

github isn't a general code Q&A site: this is a repository for the PhpSpreadsheet code library, so we provide the code and answer questions about that specific library. Sites like StackOverflow are much better places to ask general coding questions

**Reply by ZakariaBghr** [reply]
okay thanks I've made not of that
